crypto/hpke/hpke.c

```diff
@@ -0,0 +1,1390 @@
+        goto err;
+    }
+    if (ctx->suite.aead_id != OSSL_HPKE_AEAD_ID_EXPORTONLY) {
+        /* we only need nonce/key for non export AEADs */
+        ctx->noncelen = aead_info->Nn;
+        ctx->nonce = OPENSSL_malloc(ctx->noncelen);
+        if (ctx->nonce == NULL)
+            goto err;
+        if (ossl_hpke_labeled_expand(kctx, ctx->nonce, ctx->noncelen,
+                                     secret, secretlen, OSSL_HPKE_SEC51LABEL,
+                                     suitebuf, sizeof(suitebuf),
+                                     OSSL_HPKE_NONCE_LABEL,
+                                     ks_context, ks_contextlen) != 1) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        ctx->keylen = aead_info->Nk;
+        ctx->key = OPENSSL_malloc(ctx->keylen);
+        if (ctx->key == NULL)
+            goto err;
+        if (ossl_hpke_labeled_expand(kctx, ctx->key, ctx->keylen,
+                                     secret, secretlen, OSSL_HPKE_SEC51LABEL,
+                                     suitebuf, sizeof(suitebuf),
+                                     OSSL_HPKE_KEY_LABEL,
+                                     ks_context, ks_contextlen) != 1) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    }
+    ctx->exporterseclen = kdf_info->Nh;
+    ctx->exportersec = OPENSSL_malloc(ctx->exporterseclen);
+    if (ctx->exportersec == NULL)
+        goto err;
+    if (ossl_hpke_labeled_expand(kctx, ctx->exportersec, ctx->exporterseclen,
+                                 secret, secretlen, OSSL_HPKE_SEC51LABEL,
+                                 suitebuf, sizeof(suitebuf),
+                                 OSSL_HPKE_EXP_LABEL,
+                                 ks_context, ks_contextlen) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    erv = 1;
+
+err:
+    OPENSSL_cleanse(ks_context, OSSL_HPKE_MAXSIZE);
+    OPENSSL_cleanse(psk_hash, OSSL_HPKE_MAXSIZE);
+    OPENSSL_cleanse(secret, OSSL_HPKE_MAXSIZE);
+    EVP_KDF_CTX_free(kctx);
+    return erv;
+}
+
+/*
+ * externally visible functions from below here, API documentation is
+ * in doc/man3/OSSL_HPKE_CTX_new.pod to avoid duplication
+ */
+
+OSSL_HPKE_CTX *OSSL_HPKE_CTX_new(int mode, OSSL_HPKE_SUITE suite,
+                                 OSSL_LIB_CTX *libctx, const char *propq)
+{
+    OSSL_HPKE_CTX *ctx = NULL;
+
+    if (hpke_mode_check(mode) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_INVALID_ARGUMENT);
+        return NULL;
+    }
+    if (hpke_suite_check(suite) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_INVALID_ARGUMENT);
+        return NULL;
+    }
+    ctx = OPENSSL_zalloc(sizeof(OSSL_HPKE_CTX));
+    if (ctx == NULL)
+        return NULL;
+    ctx->libctx = libctx;
+    if (propq != NULL) {
+        ctx->propq = OPENSSL_strdup(propq);
+        if (ctx->propq == NULL) {
+            OSSL_HPKE_CTX_free(ctx);
```

Comment:
this call to `OSSL_HPKE_CTX_free` is useless , just call `OPENSSL_free`

Reply:
heh, you're right (think an earlier iteration did need that) - sorry for misinterpreting your earlier comment
fixed now